cql3/functions/functions.cc

```diff
@@ -463,23 +465,32 @@ function_call::contains_bind_marker() const {
 
 shared_ptr<terminal>
 function_call::make_terminal(shared_ptr<function> fun, cql3::raw_value result, cql_serialization_format sf)  {
-    if (!dynamic_pointer_cast<const collection_type_impl>(fun->return_type())) {
-        return ::make_shared<constants::value>(std::move(result));
-    }
+    static constexpr auto to_buffer = [] (const cql3::raw_value& v) {
+        if (v) {
+            return fragmented_temporary_buffer::view{bytes_view{*v}};
+        }
+        return fragmented_temporary_buffer::view{};
+    };
+
+    auto& ret_type = fun->return_type();
 
-    auto ctype = static_pointer_cast<const collection_type_impl>(fun->return_type());
-    fragmented_temporary_buffer::view res;
-    if (result) {
-        res = fragmented_temporary_buffer::view(bytes_view(*result));
+    if (ret_type->is_collection()) {
```

Comment:
This might be better as a visit() now.

Reply:
Thought so too, but then I saw `static_pointer_cast<const list_type_impl>(ret_type)`: `from_serialized` expects a `data_type`.
But I agree, so I'll add a commit that modifies `from_serialized` to take `const abstract_type&`, and then use `visit` here.

Reply:
Done